### PR TITLE
[STAGING] Introduce content graph to content

### PR DESCRIPTION
**Pull request by ilaner**
## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: link to the issue

## Description
* Introduce content graph to content
* Run create instances and prepare for mpv2 steps for id_set and mpv2
* Compare zips and send it to slack channel
 
## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Cortex XSOAR
- [ ] 6.0.0
- [ ] 6.1.0
- [ ] 6.2.0
- [ ] 6.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [ ] No

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

File: Tests/scripts/create_artifacts_graph/create_artifacts.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+                all_level_dependencies.append(dependency.content_item.object_id)
+                if dependency.is_direct:
+                    first_level_dependencies[dependency.content_item.object_id] = {
+                        "display_name": dependency.content_item.name,
+                        "mandatory": dependency.is_direct,
+                    }
+
+                    displayed_images.extend((integration.object_id for integration in pack.content_items.integration))
+            pack_dependencies[pack.object_id] = {
+                "path": str(pack.path.relative_to(Path.cwd())),
```
Comment (dorschw) [inline Tests/scripts/create_artifacts_graph/create_artifacts.py:45]: to `cwd`? not to CONTENT_PATH? or are they the same?

Comment (ilaner) [reply Tests/scripts/create_artifacts_graph/create_artifacts.py]: should be the same but changed to get_content_path to be on the safe side

File: Tests/scripts/prepare_content_packs_for_testing.sh [hunk 2]
```diff
@@ -9,6 +9,13 @@ PACK_ARTIFACTS=$ARTIFACTS_FOLDER/content_packs.zip
 EXTRACT_FOLDER=$(mktemp -d)
 ID_SET=$ARTIFACTS_FOLDER/id_set.json
 
+# check if ID_SET exists
+POSTFIX=""
```
Comment (dantavori) [inline Tests/scripts/prepare_content_packs_for_testing.sh:13]:
```suggestion
STAGING_SUFFIX=""
```

File: Tests/scripts/prepare_content_packs_for_testing.sh [hunk 3]
```diff
@@ -9,6 +9,13 @@ PACK_ARTIFACTS=$ARTIFACTS_FOLDER/content_packs.zip
 EXTRACT_FOLDER=$(mktemp -d)
 ID_SET=$ARTIFACTS_FOLDER/id_set.json
 
+# check if ID_SET exists
```
Comment (dantavori) [inline Tests/scripts/prepare_content_packs_for_testing.sh:12]:
```suggestion
# build type is staging if ID_SET doesn't exist
```

File: Tests/scripts/prepare_content_packs_for_testing.sh [hunk 4]
```diff
@@ -9,6 +9,13 @@ PACK_ARTIFACTS=$ARTIFACTS_FOLDER/content_packs.zip
...
+POSTFIX=""
+if [ ! -f "$ID_SET" ]; then
+    echo "ID_SET file not found at $ID_SET"
+    POSTFIX="_graph"
```
Comment (dantavori) [inline Tests/scripts/prepare_content_packs_for_testing.sh:16]:
```suggestion
    POSTFIX="_staging"
```

File: Tests/scripts/upload_artifacts.sh [hunk 5]
```diff
@@ -7,6 +7,15 @@ set -e
 BRANCH=${CI_COMMIT_BRANCH:-unknown}
 ARTIFACTS_DIR=${ARTIFACTS_FOLDER:-artifacts}
 
+# check if ID_SET exists
```
Comment (dantavori) [inline Tests/scripts/upload_artifacts.sh:10]: same

File: Tests/scripts/utils/compare_staging_graph.py [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+        f"Diff report for {marketplace}",
+        f'Job URL: {os.getenv("CI_JOB_URL")}',
+        f"Zip difference for {output_path}",
+    ]
+    for file in dir_cmp.common_files:
+        pack = file.removesuffix(".zip")
+        diff_files = compare_zips(zip_id_set / file, zip_graph / file, output_path / pack)
+        message.append(f'Different files for pack {pack}: {", ".join(diff_files)}')
+    if compare_indexes(index_id_set_path, index_graph_path, output_path):
+        message.append("index.json is also different")
```
Comment (dantavori) [inline Tests/scripts/utils/compare_staging_graph.py:172]:
```suggestion
        message.append("Detected differences between index.json files")
```

File: Tests/scripts/utils/compare_staging_graph.py [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+    dir_cmp.report_full_closure()
+    message = [
+        f"Diff report for {marketplace}",
+        f'Job URL: {os.getenv("CI_JOB_URL")}',
+        f"Zip difference for {output_path}",
+    ]
+    for file in dir_cmp.common_files:
+        pack = file.removesuffix(".zip")
+        diff_files = compare_zips(zip_id_set / file, zip_graph / file, output_path / pack)
+        message.append(f'Different files for pack {pack}: {", ".join(diff_files)}')
```
Comment (dantavori) [inline Tests/scripts/utils/compare_staging_graph.py:170]:
```suggestion
        if diff_files := compare_zips(zip_id_set / file, zip_graph / file, output_path / pack)
           message.append(f'Detected differences in the following files for pack {pack}: {", ".join(diff_files)}')
```

File: Tests/scripts/utils/compare_staging_graph.py [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+    if compare_indexes(index_id_set_path, index_graph_path, output_path):
+        message.append("index.json is also different")
+    if file_diff_text(output_path / "collect_tests_diff.json", collected_packs_id_set, collected_packs_graph):
+        message.append("collected tests is also different")
```
Comment (dantavori) [inline Tests/scripts/utils/compare_staging_graph.py:174]:
```suggestion
        message.append("Detected differences between collect tests results")
```

File: Tests/scripts/utils/compare_staging_graph.py [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+
+def sort_dict(dct: dict):
+    for k, v in dct.items():
+        if isinstance(v, dict):
+            sort_dict(v)
+        if isinstance(v, list):
+            try:
+                v.sort()
+            except TypeError:
+                v.sort(key=lambda item: item.get("name"))
```
Comment (dantavori) [inline Tests/scripts/utils/compare_staging_graph.py:29]: just making sure - do we always have a name field?

File: Tests/scripts/utils/compare_staging_graph.py [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+    with ZipFile(zip1, "r") as zip1_content, ZipFile(zip2, "r") as zip2_content:
+        # get the list of files in the zip files
+        zip1_content.extractall(path=zip1_files)
+        zip2_content.extractall(path=zip2_files)
+
+    # compare the directories
+    return compare_dirs(zip1_files, zip2_files, output_path)
+
+
+def compare_files(
```
Comment (dantavori) [inline Tests/scripts/utils/compare_staging_graph.py:76]: how do we manage files to skip comparison?

File: Tests/scripts/collect_tests/id_set.py [hunk 11]
```diff
@@ -3,6 +3,10 @@
 from typing import Iterable, Optional
 
 from demisto_sdk.commands.common.constants import MarketplaceVersions
+from demisto_sdk.commands.content_graph.interface.neo4j.neo4j_graph import Neo4jContentGraphInterface
```
Comment (dantavori) [inline Tests/scripts/collect_tests/id_set.py:6]: @dorschw leaving this area to you

File: Tests/Marketplace/upload_packs.py [hunk 12]
```diff
@@ -12,7 +12,7 @@
 from pathlib import Path
 
 from zipfile import ZipFile
-from typing import Any, Tuple, Union, Optional
+from typing import Any, List, Tuple, Union, Optional
```
Comment (dantavori) [inline Tests/Marketplace/upload_packs.py:15]: @yaakovpraisler please review the changes in upload flow

File: .gitlab/ci/staging.yml [hunk 13]
```diff
@@ -0,0 +1,129 @@
+graph-create-instances:
```
Comment (dantavori) [inline .gitlab/ci/staging.yml:1]: @esharf can you please have a look at the CI changes as well? (especially under `.gitlab/ci`)

**Review by dantavori — CHANGES_REQUESTED** [review]

File: Tests/Marketplace/upload_packs.py [hunk 14]
```diff
@@ -1028,7 +1027,14 @@ def main():
     install_logging('Prepare_Content_Packs_For_Testing.log', logger=logging)
     option = option_handler()
     packs_artifacts_path = option.packs_artifacts_path
-    id_set = open_id_set_file(option.id_set_path)
+    id_set = None
+    try:
+        id_set = open_id_set_file(option.id_set_path)
+    except Exception as e:
+        with Neo4jContentGraphInterface() as graph:
+            if not graph.is_graph_alive():
```
[the remark this answers is not in the document]
Comment (ilaner) [reply Tests/Marketplace/upload_packs.py:1035]: yes, removing

File: Tests/scripts/utils/compare_staging_graph.py (hunk 9 above)
Comment (ilaner) [reply Tests/scripts/utils/compare_staging_graph.py:30]: The problem here that we are comparing dictionaries with a nested list of dictionaries, and it's unsorted. We can't put dictionaries in a set, but we can sort them by a field that they always should contain.

File: Tests/scripts/utils/compare_staging_graph.py (hunk 10 above)
Comment (ilaner) [reply Tests/scripts/utils/compare_staging_graph.py:77]: added global SKIPPED_FILES

**Review by DeanArbel — COMMENTED** [review]
Looks good

**Review by dantavori — APPROVED** [review]
please add yourself as a CODEOWNER to the relevant places (gitlab ci, upload, collect tests, create artifacts)

File: Tests/Marketplace/upload_packs.py [hunk 15]
```diff
@@ -201,7 +201,7 @@ def update_index_folder(index_folder_path: str, pack_name: str, pack_path: str,
 
 
 def clean_non_existing_packs(index_folder_path: str, private_packs: list, storage_bucket: Any,
-                             storage_base_path: str, id_set: dict, marketplace: str = 'xsoar') -> bool:
+                             storage_base_path: str, pack_list: List[Pack], marketplace: str = 'xsoar') -> bool:
```
Comment (yaakovpraisler) [inline Tests/Marketplace/upload_packs.py:204]: Please fix docstring

File: Tests/Marketplace/upload_packs.py [hunk 16]
```diff
@@ -342,7 +341,7 @@ def upload_index_to_storage(index_folder_path: str, extract_destination_path: st
             # Store index.json in CircleCI artifacts
             shutil.copyfile(
                 os.path.join(index_folder_path, f'{GCPConfig.INDEX_NAME}.json'),
-                os.path.join(artifacts_dir, f'{GCPConfig.INDEX_NAME}.json'),
+                os.path.join(artifacts_dir, f'{GCPConfig.INDEX_NAME}-{"id_set" if id_set else "graph"}.json'),
```
Comment (yaakovpraisler) [inline Tests/Marketplace/upload_packs.py:344]: Is this by design or just for testing?
If its by design, then need to fix when reading this file as well.

File: Tests/Marketplace/marketplace_services.py [hunk 17]
```diff
@@ -3929,6 +3934,11 @@ def get_last_commit_from_index(service_account):
     return index_json.get('commit')
 
 
+def get_graph_entity_by_path(entity_path: Path, marketplace):
```
Comment (yaakovpraisler) [inline Tests/Marketplace/marketplace_services.py:3937]: Need to make one more method for `get_id_set_entity_by_display_name` for getting the entity by display name.

File: Tests/Marketplace/upload_packs.py (hunk 16 above)
Comment (ilaner) [reply Tests/Marketplace/upload_packs.py:344]: I'm not sure when we read this file from Gitlab artifacts, but to be on the safe side I kept the same name if we're using the id_set

**Comment by ilaner** [comment]
Decided to merge on Sunday to be safe

File: Tests/Marketplace/marketplace_services.py (hunk 17 above)
Comment (ilaner) [reply Tests/Marketplace/marketplace_services.py:3924]: Where do we use it?

Comment (yaakovpraisler) [reply Tests/Marketplace/marketplace_services.py:3924]: The name of the method was changed today, it's now `is_content_item_in_id_set` and we use it when filtering by thecontent item display name.

**Comment by ilaner** [comment]
upload flow build: https://code.pan.run/xsoar/content/-/pipelines/3982988

**Pull request merged**